sqlite/comdb2build.c

```diff
@@ -846,200 +860,260 @@ void comdb2RebuildIndex(Parse* pParse, Token* nm, Token* lnm, Token* index, int
 
 void comdb2CreateProcedure(Parse* pParse, Token* nm, Token* ver, Token* proc)
 {
+    char spname[MAX_SPNAME];
+    char sp_version[MAX_SPVERSION_LEN];
+
     Vdbe *v  = sqlite3GetVdbe(pParse);
     if (comdb2AuthenticateUserOp(v, pParse))
-        return;     
-    TokenStr(name, nm);
-    if (strlen(name) >= MAX_SPNAME) {
-        sqlite3ErrorMsg(pParse, "bad procedure name:%s", name);
+        return;
+
+    if (comdb2TokenToStr(nm, spname, sizeof(spname))) {
+        setError(pParse, SQLITE_MISUSE, "Procedure name is too long");
         return;
     }
-    struct schema_change_type* sc = new_schemachange_type();
-    strcpy(sc->table, name);
+
+    struct schema_change_type *sc = new_schemachange_type();
+    strcpy(sc->table, spname);
     sc->newcsc2 = malloc(proc->n);
     sc->addsp = 1;
+
     if (ver) {
-        TokenStr(version, ver);
-        size_t len = strlen(version);
-        if (len == 0 || len >= MAX_SPVERSION_LEN) {
-            sqlite3ErrorMsg(pParse, "bad procedure version:%s", version);
-            free_schema_change_type(sc);
-            return;
+        if (comdb2TokenToStr(ver, sp_version, sizeof(sp_version))) {
+            setError(pParse, SQLITE_MISUSE, "Procedure version is too long");
+            goto cleanup;
         }
-        strcpy(sc->fname, version);
+        strcpy(sc->fname, sp_version);
     }
+
     copyNosqlToken(v, pParse, &sc->newcsc2, proc);
     const char* colname[] = {"version"};
     const int coltype = OPFUNC_STRING_TYPE;
     OpFuncSetup stp = {1, colname, &coltype, 256};
-    comdb2prepareOpFunc(v, pParse, 1, sc, &comdb2ProcSchemaChange, (vdbeFuncArgFree)&free_schema_change_type, &stp);
+    comdb2prepareOpFunc(v, pParse, 1, sc, &comdb2ProcSchemaChange,
+                        (vdbeFuncArgFree)&free_schema_change_type, &stp);
+    return;
+
+cleanup:
+    free_schema_change_type(sc);
+    return;
 }
 
-void comdb2DefaultProcedure(Parse* pParse, Token* nm, Token* ver, int str)
+void comdb2DefaultProcedure(Parse *pParse, Token *nm, Token *ver, int str)
 {
-    Vdbe *v  = sqlite3GetVdbe(pParse);
+    char spname[MAX_SPNAME];
+    char sp_version[MAX_SPVERSION_LEN];
+
+    Vdbe *v = sqlite3GetVdbe(pParse);
+
     if (comdb2AuthenticateUserOp(v, pParse))
-        return;     
-    TokenStr(name, nm);
-    if (strlen(name) >= MAX_SPNAME) {
-        sqlite3ErrorMsg(pParse, "bad procedure name:%s", name);
+        return;
+
+    if (comdb2TokenToStr(nm, spname, sizeof(spname))) {
```

Comment:
Should this be dequoted?

Reply:
comdb2TokenToStr() also dequotes the token.